File: sentil-py/tests/test_stats.py
import pytest

import sentil
from sentil import (Formula, LiftingRegistry, NoiseModel, NoiseInteraction, SmcConfig,
                    SprtConfig, BayesConfig, SprtVerdict, BayesVerdict, SimExpr, SimModel,
                    RareEventConfig, stats)

def test_confidence_oracle_numbers():
    w = stats.wilson_interval(50, 100, 0.95)
    assert w.lower == pytest.approx(0.403831, abs=1e-5)
    assert w.upper == pytest.approx(0.596169, abs=1e-5)
    cp = stats.clopper_pearson(50, 100, 0.95)
    assert cp.lower == pytest.approx(0.398321, abs=1e-5)
    assert stats.z_score(0.95) == pytest.approx(1.95996, abs=1e-5)
    assert stats.chernoff_hoeffding_samples(0.1, 0.05) == 185

def test_noise_models():
    g = NoiseModel.gaussian(2.0, 0.5)
    assert g.mean() == pytest.approx(2.0) and g.variance() == pytest.approx(0.25)
    assert NoiseModel.cauchy(0.0, 1.0).mean() is None
    assert NoiseModel.from_json(g.to_json()).mean() == g.mean()
    res = NoiseModel.residuals([1.0, 2.0, 3.0], [1.1, 1.9, 3.2], NoiseInteraction.Additive)
    assert len(res) == 3
    assert NoiseModel.fit_gaussian([0.1, -0.2, 0.05, 0.3]).mean() is not None

def lifted():
    trace = sentil.Trace([0, 1, 2, 3], {"x": [0.5, 0.4, 0.6, 0.55]})
    lift = LiftingRegistry()
    lift.register("x", NoiseModel.gaussian(0.0, 0.2))
    return trace, lift

def test_smc_check():
    trace, lift = lifted()
    phi = Formula.parse("P>=0.5 (always (x > 0))")
    result = phi.check(trace, lift, SmcConfig(samples=500))
    assert 0.0 <= result.probability <= 1.0
    assert result.interval.lower <= result.interval.upper
    _, dist = phi.check_distribution(trace, lift, SmcConfig(samples=500))
    assert dist.count > 0 and dist.std_dev >= 0.0

def test_non_probabilistic_check_raises():
    trace, lift = lifted()
    with pytest.raises(sentil.SemanticError):
        Formula.parse("always (x > 0)").check(trace, lift)

def test_sequential_tests():
    trace, lift = lifted()
    spec = Formula.parse("P>=0.5 (always (x > 0))")
    sprt = spec.check_sequential(trace, lift, SprtConfig(0.4, 0.6, max_samples=2000))
    assert sprt.verdict in (SprtVerdict.AcceptH0, SprtVerdict.AcceptH1, SprtVerdict.Inconclusive)
    bayes = spec.check_bayesian(trace, lift, BayesConfig(0.5, max_samples=2000))
    assert bayes.verdict in (BayesVerdict.Holds, BayesVerdict.Fails, BayesVerdict.Inconclusive)

def test_sim_model_and_rare_event():
    walk = SimModel(["y"], 0.1, 16, [SimExpr.constant(0.0)],
                    [SimExpr.prev(0) + SimExpr.noise(0)], [NoiseModel.gaussian(0, 1)])
    assert len(walk.simulate(seed=7)) == 17
    system = walk.to_stochastic_system()
    phi = Formula.parse("P>=0.99 (always[0,3] (y < 5))")
    rare = phi.check_rare_event(system, RareEventConfig(particles=256))
    assert 0.0 <= rare.violation_probability <= 1.0 and rare.simulations > 0

File: sentil-cpp/tests/test_stats.cpp
#include <sentil/sentil.hpp>

#include <cmath>
#include <vector>

#include "sentil_test.hpp"

using namespace sentil;

int main() {
    auto w = stats::wilson_interval(50, 100, 0.95);
    CHECK_CLOSE(w.lower, 0.403831, 1e-5);
    CHECK_CLOSE(w.upper, 0.596169, 1e-5);
    auto cp = stats::clopper_pearson(50, 100, 0.95);
    CHECK_CLOSE(cp.lower, 0.398321, 1e-5);
    CHECK_CLOSE(cp.upper, 0.601679, 1e-5);
    CHECK_CLOSE(stats::z_score(0.95), 1.95996, 1e-5);
    CHECK(stats::chernoff_hoeffding_samples(0.1, 0.05) == 185);

    NoiseModel g = NoiseModel::gaussian(2.0, 0.5);
    CHECK_CLOSE(*g.mean(), 2.0, 1e-12);
    CHECK_CLOSE(*g.variance(), 0.25, 1e-12);
    CHECK(!NoiseModel::cauchy(0, 1).mean().has_value());
    auto residuals = NoiseModel::residuals({10, 20, 30}, {10.5, 19, 31}, NoiseInteraction::Additive);
    CHECK_CLOSE(residuals[0], 0.5, 1e-12);
    CHECK(NoiseModel::fit_gaussian({1, 2, 3, 4, 5}).mean().has_value());
    NoiseModel mix = NoiseModel::mixture({0.5, 0.5}, NoiseModel::gaussian(0, 1),
                                         NoiseModel::gaussian(5, 1));
    CHECK(g.to_json() == NoiseModel::from_json(g.to_json()).to_json());
    (void)mix;

    LiftingRegistry reg;
    reg.register_noise("x", NoiseModel::gaussian(0.0, 1.0));
    CHECK(!reg.empty() && reg.variables() == std::vector<std::string>({"x"}));
    Trace t = Trace::indexed(20);
    t.add_signal("x", std::vector<double>(20, 1.5));
    CHECK(reg.lift(t, 42).signal("x") == reg.lift(t, 42).signal("x"));

    Formula phi = Formula::parse("P>=0.5 (always (x > 0))");
    SmcResult result = phi.check(t, reg, SmcConfig{2000, 0.95, 7});
    CHECK(result.samples == 2000);
    CHECK(result.probability > 0.0 && result.probability < 1.0);
    CHECK(result.interval.lower <= result.probability && result.probability <= result.interval.upper);
    CHECK(phi.check_conservative(t, reg).samples == 10000);
    auto distribution = phi.check_distribution(t, reg, SmcConfig{500, 0.95, 1});
    CHECK(distribution.first.samples == 500 && distribution.second.count > 0);

    SprtResult sprt = phi.check_sequential(t, reg, SprtConfig{0.3, 0.7});
    CHECK(sprt.samples > 0);
    BayesResult bayes = phi.check_bayesian(t, reg, BayesConfig{0.5});
    CHECK(bayes.posterior >= 0.0 && bayes.posterior <= 1.0);

    std::vector<SimExpr> init;
    init.push_back(SimExpr::constant(0.0));
    std::vector<SimExpr> advance;
    advance.push_back(SimExpr::prev(0) * 0.5 + SimExpr::noise(0));
    std::vector<NoiseModel> noise;
    noise.push_back(NoiseModel::gaussian(0.0, 1.0));
    SimModel model({"x"}, 1.0, 20, std::move(init), std::move(advance), std::move(noise));
    CHECK(model.simulate(42).signal("x") == model.simulate(42).signal("x"));
    StochasticSystem system = model.to_stochastic_system();
    Formula rare_spec = Formula::parse("P>=0.001 (always (x < 8))");
    RareEventResult rare = rare_spec.check_rare_event(system, RareEventConfig{2048, 0.0, 1});
    CHECK(rare.simulations > 0);
    CHECK_CLOSE(rare.probability + rare.violation_probability, 1.0, 1e-9);

    return sentil_report("test_stats");
}